File: Lab 4/Lab 4/TestLearner.java
public class TestLearner {
    public static void main(String[] args) {

        Dataset d = new Dataset() {
            @Override
            public Record transform(Record r) {
                return r;
            }
        };

        d.addRecord(new Record(new Vector(2, new double[]{1, 1}), 10));
        d.addRecord(new Record(new Vector(2, new double[]{1, 2}), 13));
        d.addRecord(new Record(new Vector(2, new double[]{2, 1}), 12));
        d.addRecord(new Record(new Vector(2, new double[]{2, 2}), 15));

        Algorithm gd = new GradientDescent(0.01, 1e-9);
        SupervisedLearner sl_gd = new SupervisedLearner(d, gd);
        sl_gd.solve();
        System.out.println(sl_gd);

        Algorithm sgd = new StochasticGradientDescent(0.01, 2000, 2, 42L);
        SupervisedLearner sl_sgd = new SupervisedLearner(d, sgd);
        sl_sgd.solve();
        System.out.println(sl_sgd);

        Vector test = new Vector(2, new double[]{3, 3});
        System.out.println(sl_gd.predict(test));
        System.out.println(sl_sgd.predict(test));

        Dataset sd = d.standardize();
        Algorithm sgdStd = new StochasticGradientDescent(0.01, 2000, 2, 123L);
        SupervisedLearner sl_sgd_std = new SupervisedLearner(sd, sgdStd);
        sl_sgd_std.solve();
        System.out.println(sl_sgd_std);
        System.out.println(sl_sgd_std.predict(test));
    }
}

File: Lab 5/TestLearner.cpp
#include <iostream>
#include "GradientDescent.h"
#include "StochasticGradientDescent.h"
#include "SupervisedLearner.h"

int main() {
    Dataset d;

    d.addRecord(Record(Vector(2, {1,1}), 10));
    d.addRecord(Record(Vector(2, {1,2}), 13));
    d.addRecord(Record(Vector(2, {2,1}), 12));
    d.addRecord(Record(Vector(2, {2,2}), 15));

    GradientDescent gd(0.01, 1e-9);
    SupervisedLearner sl1(d, gd);
    sl1.solve();
    std::cout << sl1 << std::endl;

    StochasticGradientDescent sgd(0.01, 2000, 2);
    SupervisedLearner sl2(d, sgd);
    sl2.solve();
    std::cout << sl2 << std::endl;

    Vector test(2, {3,3});
    std::cout << sl1.predict(test) << std::endl;
    std::cout << sl2.predict(test) << std::endl;

    return 0;
}
